server/security.py

```diff
@@ -10,36 +10,97 @@
+cognito_eu = CognitoAuth(settings=CognitoSettings.from_global_settings(auth_settings))
+api_key_header = APIKeyHeader(name="X-API-Key", auto_error=False)
 
-logger = get_logger(__name__)
 
+def api_key_auth_required(error: bool = True):
+    """
+    Checks if a valid API key is present in the request headers.
 
-cognito_eu = CognitoAuth(settings=CognitoSettings.from_global_settings(auth_settings))
+    :param bool error: If True, raises an HTTPException if no valid API key is present, otherwise returns None.
+    """
+
+    def check_api_key(api_key: Optional[str] = Security(api_key_header)):
+        try:
+            if not api_key:
+                raise HTTPException(status_code=401, detail="API key required")
+
+            api_key_record = api_key_crud.get_by_hashed_key(api_key)
+            if (
+                not api_key_record
+                or api_key_record.revoked_at
+                or not pwd_context.verify(api_key, api_key_record.hashed_key)
+            ):
+                raise HTTPException(status_code=401, detail="Invalid API key")
+
+            return api_key_record
+        except HTTPException as e:
+            if error:
+                raise e
+            else:
+                return None
+
+    return check_api_key
 
 
-def auth_required(token: CognitoToken = Depends(cognito_eu.auth_required)):
-    if token.client_id == app_settings.AWS_COGNITO_CLIENT_ID:
-        # No need to check scopes for user tokens
-        return token
+def cognito_auth_required(error: bool = True):
+    """
+    Checks if a valid Cognito JWT token is provided in the request.
 
-    # M2M tokens: check required scope
-    if token.scope != "api":
-        raise HTTPException(status_code=401, detail="Not enough permissions")
+    :param bool error: If True, raises an HTTPException if no valid JWT token is present, otherwise returns None.
+    """
 
-    return token
+    async def check_cognito_token(req: Request):
+        try:
+            token = await cognito_eu.auth_required(req)
+
+            # check if we are dealing with a plain user token or M2M token
+            if token.client_id == app_settings.AWS_COGNITO_CLIENT_ID:
+                # No need to check scopes for user tokens
+                return token
+
+            if token.scope != "api":
+                # M2M tokens: check required scope
+                raise HTTPException(status_code=401, detail="Insufficient permissions")
+
+            return token
+        except HTTPException as e:
+            if error:
+                raise e
+            else:
+                return None
+
+    return check_cognito_token
+
+
+def auth_required(
+    token: CognitoToken | None = Depends(cognito_auth_required(error=False)),
+    api_key: APIKeyTable | None = Depends(api_key_auth_required(error=False)),
+):
+    if not token or not api_key:
+        raise HTTPException(status_code=401, detail="Insufficient permissions")
+
+    return token or api_key
```

Comment:
@acidjunk There has to be a better way to compose `Depends` similar to a logical OR, right?

Reply:
Fixed it by trying one dependency after the other, and wrapping it a new dependency that requires either the header or the JWT to be present.